### PR TITLE
NIFI-3518 Create a Morphlines processor

### PR DESCRIPTION
**Pull request by WilliamNouet**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

For all changes:

[Y] Is there a JIRA ticket associated with this PR? Is it referenced
in the commit message?

[Y] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

[Y] Has your PR been rebased against the latest commit within the target branch (typically master)?

[Y] Is your initial contribution a single, squashed commit?

For code changes:

[Y] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
[Y] Have you written or updated unit tests to verify your changes?
[N/A] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under ASF 2.0?
[N/A] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
[N/A] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
[N/A] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?
For documentation related changes:

[N/A] Have you ensured that format looks appropriate for the output in which it is rendered?
Note:

Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by joewitt** [comment]
Is there anyone familiar with morphlines that can help test/review?

**Comment by WilliamNouet** [comment]
Dear all,

First, let me introduce myself. My name is William Nouet and I am a big fan of the NiFi project. I believe this is revolutionizing the way the industry is handling data transport and transformation. I would love to be involved in the project and its open source community; so I created a JIRA (NIFI-3518) and decided to help close another one (NIFI-385).

I got some amazing feedback and guidance from the community while working on the above 2 JIRAs, updated my code to make sure it lives up to the NiFi standards. That being said, I haven’t been able to get any traction on getting the code pushed to the main branch for the past few months, hence this emailZ

Please let me know if there is anything I can do to get the ball rolling. I would be happy to re-work on my code if you think it can be improved and would more than happy to meet in person in a bar or a coffee shop if some of you guys are in the NYC metro area.

Thank you in advance.

Best,

William

Envoyé de mon iPhone

**Comment by joewitt** [comment]
@WilliamNouet Sorry for the trouble.  We try to have a good review model for these things and it seems like we just dont have anyone familiar with morphlines that can help.  I am happy to help review it and work with you to get it in though.  Can you please rebase, change nifi version references to 1.5.0-SNAPSHOT, and we can go from there.  I'm almost certain there will be a lot of LICENSE and NOTICE updates required which I dont see yet.

**Comment by joewitt** [comment]
@WilliamNouet  can you please see my last set of comments.  A rebase and some progress on the licensing will be necessary. I can definitely help that.  I have some time over the next couple of days and this is one I'd be happy to try and knock out if time allows.

**Comment by WilliamNouet** [comment]
@joewitt Thanks for getting back to me, and sorry for the delay; I was away without access to a laptop. I just pushed a commit with an updated snapshot version and licensing.

**Comment by WilliamNouet** [comment]
@joewitt, did you get a chance to review the latest commit?

**Comment by joewitt** [comment]
Bill,

Some feedback after reviewing in more detail.

1. There are no unit tests.  I can see why they'd be non-portable given it is executing a process but it does seem like a good candidate for an integration-test at least.  We should definitely have something.  We do have a naming convention for integration tests to execute.
2. The nar is not connected to the nifi-nar-bundles pom
3. The nar is not connected to the nifi-assembly.  This is perfectly fine though if you dont want the nar to be available in the build by default but rather something someone opts into.  I'm ok either way just pointing this out.
4. The L&N updates within the nar are a good start but are incomplete with regard to the totality of the jars present within the nar.
5. The L&N updates in the nar will also need to be validated to see if any need to be carried forward into the nifi-assembly nar.
6. There are unused imports which cause checkstyle/rat failures which is why travis failed on all builds.
7. I'm not an expert on morphlines but it seems like it could be used to do dangerous things on the filesystem.  We have a RestrictedComponent annotation for this purpose so it is probably best to add that to the processor.
8. The processor name for a 'users point of view' should probably be 'ExecuteMorphline'
9. You'll probably want to provide some documentation/additional documentation as many other processors do helping the user (and reviewer to evaluate).
10. Can you update to latest nifi version which is 1.6.0-SNAPSHOT in your next commit.  Not a big deal as I can do that easily enough but it helps.

This looks like a lot but for the L&N we have tons of examples.  The rest are pretty straightforward to sort out.

Let me know if you plan to tackle these.

Thanks

**Comment by joewitt** [comment]
i added a file to the JIRA which is a patch you can apply for the simple pom updates for version

**Comment by WilliamNouet** [comment]
@joewitt thanks for the feedback! That was quite a bit of changes to make on my side, but hopefully the latest commit captures everything. I will need to work on the rebase a bit later though.

**Comment by binhnv** [comment]
@WilliamNouet there is already a `nifi-kite-bundle` which has `kite`'s dependencies, does it make sense to move this processor to that bundle?

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ExecuteMorphline.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if ( flowFile == null ) {
+            return;
+        }
+
+        FlowFile originalFlowFile = session.clone(flowFile);
+        final AtomicLong written = new AtomicLong(0L);
+        final byte[] value = new byte[(int) flowFile.getSize()];
```
Comment (binhnv) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ExecuteMorphline.java]: This is not going to work for big file

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ExecuteMorphline.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+            getLogger().error("Error while processing the flowFile through Morphlines");
+            session.transfer(flowFile, REL_ORIGINAL);
+            session.transfer(originalFlowFile, REL_FAILURE);
+            Notifications.notifyRollbackTransaction(morphline);
+            morphlineContext.getExceptionHandler().handleException(e, null);
+        }
+        Notifications.notifyShutdown(morphline);
+    }
+
+    private static final class Collector implements Command {
```
Comment (binhnv) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ExecuteMorphline.java]: The same as above, this implementation is not going to work with big input file

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ExecuteMorphline.java (hunk 1 above)
Comment (WilliamNouet) [reply nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ExecuteMorphline.java]: It depends on your definition of "big" files. I believe that is not in the user interest to have big files (> 1 GB) being processed through the data flow as most processors under perform with this kind of files. So I do not think it is an issue here; I have seen the processor getting tremendous results on files up to 0.5 GB.

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ExecuteMorphline.java (hunk 2 above)
Comment (WilliamNouet) [reply nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ExecuteMorphline.java]: Same answer as the above.

**Comment by WilliamNouet** [comment]
@joewitt & @binhnv would you have an update on the above recent comments and modifications?
Thanks!

**Comment by MikeThomsen** [comment]
@WilliamNouet I did some searching, and it appears that Morphline is abandoned. The last update to the Kitesdk morphline package is in 2015. I'm going to close this out because we can't accept submissions that are based on abandoned core dependencies. If I missed something and Morphline is still active, feel free to reopen. Regardless, thanks for the contribution. It was a solid effort.

**Pull request closed without merging**